simSPI/linear_simulator/modules.py

```diff
@@ -0,0 +1,397 @@
+            :, None, :, :
+        ]  # add a dummy channel (for consistency w/ img fmt)
+        # --> B,C,H,W
+        return projection
+
+
+"""Module that multiplies ctf to fourier transform of the projection."""
+
+
+class CTF(torch.nn.Module):
+    """Class that multiplies ctf to fourier transform of the projection.
+
+    Parameters
+    ----------
+    config: class
+        Class containing parameters of the CTF
+
+    """
+
+    def __init__(self, config):
+        super(CTF, self).__init__()
+
+        self.config = config
+        self.wavelength = self._get_ewavelength()
+        self.frequency = 1.0 / (self.config.ctf_size * self.config.pixel_size)
+        n2 = float(self.config.ctf_size // 2)
+        ax = torch.arange(-n2, n2 + self.config.ctf_size % 2)
+        mx, my = torch.meshgrid(ax, ax)
+        self.register_buffer("r2", mx ** 2 + my ** 2)
+        self.register_buffer("r", torch.sqrt(self.r2))
+        self.register_buffer("angleFrequency", torch.atan2(my, mx))
+
+    def _get_ewavelength(self):
+        """Output the wavelength of the electron beam.
+
+        Returns
+        -------
+        wavelength: float
+            wavelength of the electron beam
+        """
+        wavelength = 12.2639 / np.sqrt(
+            self.config.kV * 1e3 + 0.97845 * self.config.kV ** 2
+        )
+        return wavelength
+
+    def get_ctf(self, ctf_params):
+        """Output the CTF for a given params.
+
+        Parameters
+        ----------
+        ctf_params: dict of type str to {tensor}
+            contains "defocusU", "defocusV", "defocusAngle",
+            all of shape (chunks,1,1,1)
+
+        Returns
+        -------
+        hFourier: torch.Tensor
+            ctf of shape (chunks,1,ctf_size,ctf_size)
+        """
+        defocusU = ctf_params["defocusU"]
+        defocusV = ctf_params["defocusV"]
+        angleAstigmatism = ctf_params["defocusAngle"]
+        elliptical = (
+            defocusV * self.r2
+            + (defocusU - defocusV)
+            * self.r2
+            * torch.cos(self.angleFrequency - angleAstigmatism) ** 2
+        )
+        defocusContribution = (
+            np.pi * self.wavelength * 1e4 * elliptical * self.frequency ** 2
+        )
+        abberationContribution = (
+            -np.pi
+            / 2.0
+            * self.config.cs
+            * (self.wavelength ** 3)
+            * 1e7
+            * self.frequency ** 4
+            * self.r2 ** 2
+        )
+
+        argument = abberationContribution + defocusContribution
+
+        hFourier = (1 - self.config.amplitude_contrast ** 2) ** 0.5 * torch.sin(
+            argument
+        ) + self.config.amplitude_contrast * torch.cos(argument)
+
+        if self.config.bfactor == 0:
+            decay = (
+                np.sqrt(-np.log(self.config.valueNyquist))
+                * 2.0
+                * self.config.pixel_size
+            )
+            envelope = torch.exp(-self.frequency ** 2 * decay ** 2 * self.r2)
```

Comment:
can you put in some documentation as to why this is a good default choice? where did it come from?

Reply:
like what do we need (prevent aliasing? needs to go to zero at Nyquist or not? etc)

Reply:
Good questions @geoffwoollard , I suggest someone opens an issue to discuss CTF and get to the bottom of this. There are several possible strategies and they all need to be explained and properly argumented.

Reply:
@geoffwoollard  Not sure what you mean by default choice. This piece of code sees if the bfactor of the envelope is 0 or not. If it is not zero then bfactor is used to create the envelope using the conventional formula. If it is zero then the value_nyquist is used which basically is given by user to say what is value of the envelope they want at nyquist frequency (see https://github.com/harshit-gupta-cor/simSPI/blob/004cbbda15fac64a721501cd9f41b5fe94553882/simSPI/linear_simulator/configs/betagal_distribution.cfg#L40). This is a more intuitive thing to give than bfactor. 

I agree this needs more documentation in the docstring which I will do. I don't think the value of bfactor determines aliasing. The latter I believe depends on the number of the samples we take of CTF between 0 and nyquist frequency (size of the CTF).

Reply:
https://github.com/compSPI/simSPI/issues/21